LoadError: libffi.so.6: cannot open shared object file

## 背景

https://app.circleci.com/pipelines/github/iwaseasahi/christchurches-map/1429/workflows/69e6fe0f-d19b-4187-9b30-cf5e92e77639/jobs/2751

circleci の build に失敗するようになりました。
修正します。

## 目的

circleci の build を成功させる

## 完成条件

* 原因を調査すること
* 解決すること

```
fromRequireContextWithGlobalFallback.js:21 Error: Module build failed (from ./node_modules/babel-loader/lib/index.js):
SyntaxError: /usr/src/app/app/javascript/components/like_rankings/Church.js: Support for the experimental syntax 'jsx' isn't currently enabled (9:17):

   7 |     let urlPath;
   8 |     if (url) {
>  9 |       urlPath = <p><a target="_blank" href={url}>ホームページはこちら</a></p>
     |                 ^
  10 |     }
  11 |
  12 |     return (

Add @babel/preset-react (https://git.io/JfeDR) to the 'presets' section of your Babel config to enable transformation.
If you want to leave it as-is, add @babel/plugin-syntax-jsx (https://git.io/vb4yA) to the 'plugins' section to enable parsing.
    at Parser._raise (usr/src/app/node_modules/@babel/parser/lib/index.js:798)
    at Parser.raiseWithData (usr/src/app/node_modules/@babel/parser/lib/index.js:791)
    at Parser.expectOnePlugin (usr/src/app/node_modules/@babel/parser/lib/index.js:3272)
    at Parser.parseExprAtom (usr/src/app/node_modules/@babel/parser/lib/index.js:11514)
    at Parser.parseExprSubscripts (usr/src/app/node_modules/@babel/parser/lib/index.js:11081)
    at Parser.parseUpdate (usr/src/app/node_modules/@babel/parser/lib/index.js:11061)
    at Parser.parseMaybeUnary (usr/src/app/node_modules/@babel/parser/lib/index.js:11039)
    at Parser.parseExprOps (usr/src/app/node_modules/@babel/parser/lib/index.js:10882)
    at Parser.parseMaybeConditional (usr/src/app/node_modules/@babel/parser/lib/index.js:10856)
    at Parser.parseMaybeAssign (usr/src/app/node_modules/@babel/parser/lib/index.js:10814)
    at Parser.parseMaybeAssign (usr/src/app/node_modules/@babel/parser/lib/index.js:10843)
    at Parser.parseExpressionBase (usr/src/app/node_modules/@babel/parser/lib/index.js:10754)
    at usr/src/app/node_modules/@babel/parser/lib/index.js:10748
    at Parser.allowInAnd (usr/src/app/node_modules/@babel/parser/lib/index.js:12589)
    at Parser.parseExpression (usr/src/app/node_modules/@babel/parser/lib/index.js:10748)
    at Parser.parseStatementContent (usr/src/app/node_modules/@babel/parser/lib/index.js:12927)
    at Parser.parseStatement (usr/src/app/node_modules/@babel/parser/lib/index.js:12796)
    at Parser.parseBlockOrModuleBlockBody (usr/src/app/node_modules/@babel/parser/lib/index.js:13385)
    at Parser.parseBlockBody (usr/src/app/node_modules/@babel/parser/lib/index.js:13376)
    at Parser.parseBlock (usr/src/app/node_modules/@babel/parser/lib/index.js:13360)
    at Parser.parseStatementContent (usr/src/app/node_modules/@babel/parser/lib/index.js:12872)
    at Parser.parseStatement (usr/src/app/node_modules/@babel/parser/lib/index.js:12796)
    at Parser.parseIfStatement (usr/src/app/node_modules/@babel/parser/lib/index.js:13159)
    at Parser.parseStatementContent (usr/src/app/node_modules/@babel/parser/lib/index.js:12841)
    at Parser.parseStatement (usr/src/app/node_modules/@babel/parser/lib/index.js:12796)
    at Parser.parseBlockOrModuleBlockBody (usr/src/app/node_modules/@babel/parser/lib/index.js:13385)
    at Parser.parseBlockBody (usr/src/app/node_modules/@babel/parser/lib/index.js:13376)
    at Parser.parseBlock (usr/src/app/node_modules/@babel/parser/lib/index.js:13360)
    at Parser.parseFunctionBody (usr/src/app/node_modules/@babel/parser/lib/index.js:12245)
    at Parser.parseFunctionBodyAndFinish (usr/src/app/node_modules/@babel/parser/lib/index.js:12229)
    at Parser.parseMethod (usr/src/app/node_modules/@babel/parser/lib/index.js:12179)
    at Parser.pushClassMethod (usr/src/app/node_modules/@babel/parser/lib/index.js:13831)
    at Parser.parseClassMemberWithIsStatic (usr/src/app/node_modules/@babel/parser/lib/index.js:13721)
    at Parser.parseClassMember (usr/src/app/node_modules/@babel/parser/lib/index.js:13659)
    at usr/src/app/node_modules/@babel/parser/lib/index.js:13603
    at Parser.withSmartMixTopicForbiddingContext (usr/src/app/node_modules/@babel/parser/lib/index.js:12566)
    at Object.<anonymous> (application-9d3ea54cad0b7a1c14af.js:2)
    at n (bootstrap:19)
    at l (.*$:9)
    at fromRequireContext.js:13
    at Object.getConstructor (fromRequireContextWithGlobalFallback.js:15)
    at Object.mountComponents (index.js:89)
    at HTMLDocument.handleMount (index.js:149)
    at Object.e.dispatch (turbolinks.source.js:5)
    at r.notifyApplicationAfterPageLoad (turbolinks.source.js:6)
    at r.visitCompleted (turbolinks.source.js:6)
(anonymous) @ fromRequireContextWithGlobalFallback.js:21
fromRequireContextWithGlobalFallback.js:22 ReferenceError: like_rankings is not defined
    at eval (eval at module.exports (fromGlobal.js:12), <anonymous>:1:1)
    at module.exports (fromGlobal.js:12)
    at Object.getConstructor (fromRequireContextWithGlobalFallback.js:19)
    at Object.mountComponents (index.js:89)
    at HTMLDocument.handleMount (index.js:149)
    at Object.e.dispatch (turbolinks.source.js:5)
    at r.notifyApplicationAfterPageLoad (turbolinks.source.js:6)
    at r.visitCompleted (turbolinks.source.js:6)
    at r.complete (turbolinks.source.js:5)
    at r.<anonymous> (turbolinks.source.js:5)
(anonymous) @ fromRequireContextWithGlobalFallback.js:22
index.js:100 [react-rails] Cannot find component: 'like_rankings/Church' for element <div data-react-class=​"like_rankings/​Church" data-react-props=​"{"likesCount":​1,"name":​"新宿シャローム教会","groupName":​"ニューホープ・インターナショナル・アジア・ジャパン","address":​"東京都新宿区百人町1-23-24 アミューズ大久保ビル2F","churchPath":​"/​churches/​2324","url":​"http:​/​/​shinjuku-shalom.com/​"}​" data-react-cache-id=​"like_rankings/​Church-0">​</div>​
index.js:103 Uncaught Error: Cannot find component: 'like_rankings/Church'. Make sure your component is available to render.
    at Object.mountComponents (index.js:103)
    at HTMLDocument.handleMount (index.js:149)
    at Object.e.dispatch (turbolinks.source.js:5)
    at r.notifyApplicationAfterPageLoad (turbolinks.source.js:6)
    at r.visitCompleted (turbolinks.source.js:6)
    at r.complete (turbolinks.source.js:5)
    at r.<anonymous> (turbolinks.source.js:5)
    at turbolinks.source.js:5
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -55,6 +55,10 @@ jobs:
           name: Wait for DB
           command: dockerize -wait tcp://127.0.0.1:3306 -timeout 120s
       - run:
+          name: sudo ln -s libffi
+          command: |
+            sudo ln -s /usr/lib/x86_64-linux-gnu/libffi.so.7.1.0 /usr/lib/x86_64-linux-gnu/libffi.so.6
+      - run:
           name: Database setup
           command: bundle exec rake db:create db:migrate
   test:

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,7 +1,7 @@
 version: 2.1
 
 ruby_image: &ruby_image
-  circleci/ruby:3.0-node-browsers-legacy
+  circleci/ruby:3.0.2-node-browsers-legacy
 
 executors:
   default_executor:

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -33,11 +33,11 @@ jobs:
       - run: bundle -v
       - restore_cache:
           keys:
-            - bundle-v2-{{ checksum "Gemfile.lock" }}
-            - bundle-v2-
+            - bundle-v3-{{ checksum "Gemfile.lock" }}
+            - bundle-v3-
       - run: bundle check || bundle install --clean
       - save_cache:
-          key: bundle-v2-{{ checksum "Gemfile.lock" }}
+          key: bundle-v3-{{ checksum "Gemfile.lock" }}
           paths:
             - vendor/bundle
       - restore_cache:
@@ -67,8 +67,8 @@ jobs:
       - checkout
       - restore_cache:
           keys:
-            - bundle-v2-{{ checksum "Gemfile.lock" }}
-            - bundle-v2-
+            - bundle-v3-{{ checksum "Gemfile.lock" }}
+            - bundle-v3-
       - run: bundle
       - run:
           name: Wait for DB
@@ -107,8 +107,8 @@ jobs:
       - checkout
       - restore_cache:
           keys:
-            - bundle-v2-{{ checksum "Gemfile.lock" }}
-            - bundle-v2-
+            - bundle-v3-{{ checksum "Gemfile.lock" }}
+            - bundle-v3-
       - run: bundle
       - run:
           name: Wait for DB
@@ -150,8 +150,8 @@ jobs:
             - "14:a4:f5:51:a4:70:79:7b:5b:63:d0:d7:e9:24:e2:d7"
       - restore_cache:
           keys:
-            - bundle-v2-{{ checksum "Gemfile.lock" }}
-            - bundle-v2-
+            - bundle-v3-{{ checksum "Gemfile.lock" }}
+            - bundle-v3-
       - run: bundle
       - run:
           name: Deploy Check to AWS
@@ -167,8 +167,8 @@ jobs:
             - "14:a4:f5:51:a4:70:79:7b:5b:63:d0:d7:e9:24:e2:d7"
       - restore_cache:
           keys:
-            - bundle-v2-{{ checksum "Gemfile.lock" }}
-            - bundle-v2-
+            - bundle-v3-{{ checksum "Gemfile.lock" }}
+            - bundle-v3-
       - run: bundle
       - run:
           name: Deploy to AWS

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -55,10 +55,6 @@ jobs:
           name: Wait for DB
           command: dockerize -wait tcp://127.0.0.1:3306 -timeout 120s
       - run:
-          name: sudo ln -s libffi
-          command: |
-            sudo ln -s /usr/lib/x86_64-linux-gnu/libffi.so.7.1.0 /usr/lib/x86_64-linux-gnu/libffi.so.6
-      - run:
           name: Database setup
           command: bundle exec rake db:create db:migrate
   test:

--- a/babel.config.js
+++ b/babel.config.js
@@ -22,8 +22,10 @@ module.exports = function(api) {
         {
           targets: {
             node: 'current'
-          }
-        }
+          },
+          modules: 'commonjs'
+        },
+        '@babel/preset-react'
       ],
       (isProductionEnv || isDevelopmentEnv) && [
         '@babel/preset-env',
@@ -33,6 +35,13 @@ module.exports = function(api) {
           corejs: 3,
           modules: false,
           exclude: ['transform-typeof-symbol']
+        }
+      ],
+      [
+        '@babel/preset-react',
+        {
+          development: isDevelopmentEnv || isTestEnv,
+          useBuiltIns: true
         }
       ]
     ].filter(Boolean),
@@ -68,13 +77,21 @@ module.exports = function(api) {
       [
         '@babel/plugin-transform-runtime',
         {
-          helpers: false
+          helpers: false,
+          regenerator: true,
+          corejs: false
         }
       ],
       [
         '@babel/plugin-transform-regenerator',
         {
           async: false
+        }
+      ],
+      isProductionEnv && [
+        'babel-plugin-transform-react-remove-prop-types',
+        {
+          removeImport: true
         }
       ]
     ].filter(Boolean)

--- a/package.json
+++ b/package.json
@@ -1,11 +1,11 @@
 {
   "dependencies": {
-    "@babel/preset-react": "^7.10.1",
+    "@babel/preset-react": "^7.14.5",
     "@rails/webpacker": "5.4.2",
     "babel-plugin-transform-react-remove-prop-types": "^0.4.24",
     "prop-types": "^15.7.2",
-    "react": "^16.13.1",
-    "react-dom": "^16.13.1",
+    "react": "^17.0.2",
+    "react-dom": "^17.0.2",
     "react_ujs": "^2.6.1",
     "webpack": "^4.46.0",
     "webpack-cli": "^3.3.12"

